modules/common/src/main/java/com/opengamma/sdk/common/v3/ServiceInvokerBuilder.java

```diff
@@ -0,0 +1,316 @@
+      t.setName("ServiceInvoker-" + t.getName());
+      t.setDaemon(true);
+      return t;
+    };
+    ScheduledExecutorService executor =
+        Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors(), threadFactory);
+    return executor;
+  }
+
+  //-------------------------------------------------------------------------
+  // an interceptor that performs basic logging
+  private static class LoggingInterceptor implements Interceptor {
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+      Request request = chain.request();
+      log.debug("Call service: {} {}", request.method(), request.url());
+      Response response = chain.proceed(request);
+      log.debug("Service responded: {}", response.code());
+      return response;
+    }
+  }
+
+  //An interceptor that adds the User-Agent header and exposes useful information about the SDK and runtime.
+  private static class UserAgentHeaderInterceptor implements Interceptor {
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+      Request initialRequest = chain.request();
+      Request modifiedRequest = initialRequest.newBuilder()
+          .header("User-Agent", USER_AGENT)
+          .build();
+      return chain.proceed(modifiedRequest);
+    }
+  }
+
+  // an interceptor that adds the access token, and refreshes it when necessary
+  private static class TokenInterceptor implements Interceptor {
+    /** The lock protecting the token. */
+    private final Lock lock = new ReentrantLock();
+    /** The auth client. */
+    private volatile AuthClient authClient;
+    /** The current token. */
+    private volatile AccessTokenResult token;
+  
+    // initializes the state, to ensure that ServiceInvoker is pure immutable wrt Java Memory Model
+    void init(AuthClient authClient, Credentials credentials) {
+      this.authClient = authClient;
+      token = credentials.authenticate(authClient);
+    }
+  
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+      // do nothing for auth
+      Request initialRequest = chain.request();
+      if (initialRequest.url().pathSegments().contains("auth")) {
+        return chain.proceed(initialRequest);
+      }
+  
+      // try using the current access token, unless not present
+      AccessTokenResult copyOfToken = token;
+      if (copyOfToken != null) {
+        log.trace("Add token: {}", copyOfToken.getAccessToken());
+        Request modifiedRequest = initialRequest.newBuilder()
+            .header(AUTHORIZATION, "Bearer " + copyOfToken.getAccessToken())
+            .build();
+        Response response = chain.proceed(modifiedRequest);
+        if (response.code() != 401) {
+          return response;
+        }
+      }
+  
+      // try to get a new token
+      lock.lock();
+      try {
+        if (token == null) {
+          throw new AuthenticationException("Authentication failed: Unable to retry");
```

Comment:
According to the current version of AuthenticationException on the `version/3.x` branch, I don't think this would work, as the constructor requires 2/3 parameters.

https://github.com/OpenGamma/JavaSDK/blob/version/3.x/modules/common/src/main/java/com/opengamma/sdk/common/auth/v3/AuthenticationException.java